Split mockwebserver package

yschimke opened the pull request:
Create a mockwebserver class without any junit dependencies. I think this might be viable commit to unblock graal testing work.

yschimke commented on okhttp-mockwebserver/README.md:
````diff
@@ -0,0 +1,160 @@
+### Download
+
+```kotlin
+testImplementation("com.squareup.okhttp3:mockwebserver:4.8.0")
````
TODO

yschimke commented on okhttp-mockwebserver/src/main/kotlin/okhttp3/mockwebserver/internal/duplex/MockWebServerAsserts.kt:
```diff
@@ -0,0 +1,31 @@
+package okhttp3.mockwebserver.internal.duplex
+
+open class MockWebServerAsserts {
```
@swankjesse McDonald's suggestion for avoiding any dependencies inside of MockWebServer module

yschimke commented:
PR is on top of commits from https://github.com/square/okhttp/pull/6192

yschimke commented:
@swankjesse Can I eventually squash merge this without taking credit for @tokuhirom work? Or should I histedit to have a clean set of my commits on top of his PR commit?

yschimke commented on mockwebserver/build.gradle:
```diff
@@ -22,17 +23,17 @@ afterEvaluate { project ->
-  ]
-}
-check.dependsOn(japicmp)
+//task japicmp(type: me.champeau.gradle.japicmp.JapicmpTask, dependsOn: 'jar') {
```
I'm not sure how to keep this working?

```

* What went wrong:
Execution failed for task ':mockwebserver:japicmp'.
> A failure occurred while executing JApicmp check comparing [mockwebserver-4.10.0-SNAPSHOT.jar] with [mockwebserver-3.14.1.jar]
   > Detected binary changes between mockwebserver-4.10.0-SNAPSHOT.jar and mockwebserver-3.14.1.jar. See failure report at file:///home/circleci/project/mockwebserver/build/reports/japi.txt
```

swankjesse reviewed (commented):
I don’t like how the names work. Explaining that `okhttp-mockwebserver` is new and `mockwebserver` is old is confusing and arbitrary.

I’ve been thinking about this all weekend and I’ve got some ideas, though nothing particularly satisfying. Some ideas:

 * OkHttpServer – we’d do a potentially huge amount of work to do a ‘v2’ of the API for this thing, focused on the server role. I don’t really like this because the HTTP API space is overcrowded. We could potentially differentiate by focusing on use cases like serverless and graal. But I don’t think it’s a space that needs more participants.

 * MockWebServerRule – break API compatibility and move the JUnit rule into a new class, and simply stop implementing the Rule interface in MockWebServer. I hate breaking API compatibility.

 * MockWebServer2 – probably the most honest. It’s ugly, but to readers it’s completely unambiguous which one is the good one. Could also use another number, like 3 to go alongside okhttp3.

I think MockWebServer2 is the least bad here. Probably something like this:

### com.squareup.okhttp3:mockwebserver2:4.11.0
```
class MockWebServer2 : Closeable {
  ... all the same APIs as today except JUnit dependency
}
```

### com.squareup.okhttp3:mockwebserver:4.11.0
```
@Deprecated("Replaced with MockWebServer2...")
class MockWebServer : ExternalResource, Closeable {
  val delegate = MockWebServer2()
  ... all the same APIs as today, includingJUnit dependency
}
```

I’m grumpy about JUnit breaking compatibility between 4.x and 5.x. We wouldn’t need to do this if they hadn’t! But I don’t want to repeat their mistake by breaking compatibility ourselves.

swankjesse commented:
Also I think we probably change packages from `package okhttp3.mockwebserver` to `package mockwebserver2`.

JakeWharton commented:
I prefer 1 or 2. For 2, we can release a version with both and print if you use the wrong one. Then eventually remove the rule from the old one.

swankjesse commented:
Oh, one more: if we change the package name from `okhttp3.mockwebserver` to just `mockwebserver2`, we can keep the nice class name, `MockWebServer`.

yschimke commented:
I'm in favour of 3 and eventually 1.  For now, a compatible version 2 without Junit.  And longer term a top level project for a MockWebServer. I'll work on that since I think it's blocking testing with graal, or at least I hope it's only junit4 vintage runner that needs the serialization APIs.

yschimke commented:
False alert? It's now inherited from the superclass.

```
Comparing binary compatibility of  against 
***! MODIFIED CLASS: PUBLIC FINAL okhttp3.mockwebserver.MockWebServer  (not serializable)
	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
	---! REMOVED INTERFACE: java.io.Closeable
	---! REMOVED INTERFACE: java.lang.AutoCloseable
```

yschimke commented:
> Comparing binary compatibility of  against 
> ***! MODIFIED CLASS: PUBLIC FINAL okhttp3.mockwebserver.MockWebServer  (not serializable)
> 	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
> 	---! REMOVED INTERFACE: java.io.Closeable
> 	---! REMOVED INTERFACE: java.lang.AutoCloseable

https://github.com/siom79/japicmp/issues/271

adriancole commented:
option 2 sgtm it is more intuitive even if someone didn't read this thread. thanks!

swankjesse commented on mockwebserver/build.gradle:
```diff
@@ -8,7 +8,7 @@ jar {
 dependencies {
   api project(':okhttp')
-  api deps.junit
+  compileOnly deps.junit
```
I think I’d prefer we keep this as-is to preserve compatibility, and introduce a new module that has all the code

yschimke replied:
I think this is actually cleaner.  You should only be affected after an upgrade of dependencies, and when your tests no longer run because you are missing junit the fix is trivial and obvious.  But critically, it avoids us having the confusion of an additional module.

yschimke replied:
If this change popped up at runtime of the app, I'd 100% agree with you.  What module would you suggest if we split it out?

swankjesse commented on mockwebserver/src/main/kotlin/mockwebserver/MockWebServer.kt:
```diff
@@ -0,0 +1,1149 @@
+ * limitations under the License.
+ */
+
+package mockwebserver
```
I think we want mockwebserver2 or mockwebserver3. It’s definitely uglier! But I hate trying to remember whether org.junit or junit.framework is the right one, and this is similarly ambiguous.

yschimke replied:
I actually like it.  mockwebserver3 matches okhttp3.

swankjesse commented on mockwebserver/src/main/kotlin/okhttp3/mockwebserver/MockWebServer.kt:
```diff
@@ -17,1144 +17,53 @@
-      }
-      dispatcher.shutdown()
-    }
+class MockWebServer : TestRule, MockWebServer() {
```
nice.

swankjesse replied:
wanna delegate instead of subclassing? It’s a bit more boilerplate but then we don’t need to open ’er up.

swankjesse commented on mockwebserver/src/main/kotlin/okhttp3/mockwebserver/internal/duplex/MockWebServerAsserts.kt:
```diff
@@ -0,0 +1,31 @@
+package okhttp3.mockwebserver.internal.duplex
+
+open class MockWebServerAsserts {
```
open?!

swankjesse replied:
Totally not understanding why this is a class and not an object

yschimke replied:
It's an override mechanism.  The other implementation delegates to junit

yschimke replied:
From the original comment, happy to find alternate design that by default works sensibly, but allows reverting back to using the junit API methods by override.

swankjesse commented on samples/guide/build.gradle:
```diff
@@ -2,6 +2,7 @@ apply plugin: 'kotlin-kapt'
 
 dependencies {
   implementation project(':okhttp')
+  implementation deps.junit
```
I’d prefer a solution that doesn’t require downstream projects to change their Gradle configs!

swankjesse commented:
Another take on this problem: https://github.com/square/okhttp/pull/6326
